api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto

```diff
@@ -70,6 +72,15 @@ message Config {
     // How is the value encoded, default is NONE (not encoded).
     // The value will be decoded accordingly before storing to metadata.
     ValueEncode encode = 5;
+
+    // If present, the header's value will be matched and substituted with this. If
+    // there is no match, no metadata is added.
+    //
+    // This is only used for `on_header_present`, where substitutions will be applied
+    // if a match with the header's value happened.
+    //
+    // Note: if the `value` field is non-empty this field should be empty.
+    type.matcher.v3.RegexMatchAndSubstitute regex_value_rewrite = 6;
```

Comment:
You might add `(udpa.annotations.field_migrate).oneof_promotion` annotations for this and value.